### PR TITLE
AN-1018 Implement Bindings

### PR DESCRIPTION
**Pull request by xiazcy**
*Issue #, if available:*
AN-1018

*Description of changes:*
Implemented bindings in bytecode and string requests. Commented out GA workflows for Tinkerpop test suite for now. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: gremlin-go/driver/connection_test.go [hunk 1]
```diff
@@ -233,7 +233,7 @@ func TestConnection(t *testing.T) {
 
 	// No authentication integration test with graphs loaded and alias configured server
 	testNoAuthWithAliasUrl := getEnvOrDefaultString("GREMLIN_SERVER_URL", "ws://localhost:8182/gremlin")
-	testNoAuthWithAliasEnable := getEnvOrDefaultBool("RUN_INTEGRATION_WITH_ALIAS_TESTS", false)
```
Comment (lyndonbauto) [inline gremlin-go/driver/connection_test.go]: Is this supposed to be checked in at this point?

File: gremlin-go/driver/graphBinary.go [hunk 2]
```diff
@@ -621,6 +624,7 @@ func vertexPropertyReader(buffer *bytes.Buffer, typeSerializer *graphBinaryTypeS
 }
 
 //Format: {Labels}{Objects}
+// TODO: Path serialization is currently incomplete as Labels are represented as list of lists due to lack of native set types in go. Fully functional Path serialization will be implemented when set is implemented in AN-1032
```
Comment (lyndonbauto) [inline gremlin-go/driver/graphBinary.go]: I thought this was fixed with our path type? Are our paths incorrect?

File: gremlin-go/driver/graphBinary.go [hunk 3]
```diff
@@ -845,7 +849,7 @@ func bulkSetReader(buffer *bytes.Buffer, typeSerializer *graphBinaryTypeSerializ
 }
 
 // Format: a single string representing the enum value
-func enumReader(buffer *bytes.Buffer, _ *graphBinaryTypeSerializer) (interface{}, error) {
+func enumReader(buffer *bytes.Buffer, typeSerializer *graphBinaryTypeSerializer) (interface{}, error) {
```
Comment (lyndonbauto) [inline gremlin-go/driver/graphBinary.go]: ?

File: gremlin-go/driver/graphBinary.go [hunk 4]
```diff
@@ -983,6 +1028,8 @@ func (serializer *graphBinaryTypeSerializer) getSerializerToWrite(val interface{
 		return &graphBinaryTypeSerializer{dataType: PType, writer: pWriter, logHandler: serializer.logHandler}, nil
 	case textP, TextPredicate:
 		return &graphBinaryTypeSerializer{dataType: TextPType, writer: textPWriter, logHandler: serializer.logHandler}, nil
+	case *Binding:
```
Comment (lyndonbauto) [inline gremlin-go/driver/graphBinary.go]: If you have a pointer and non-pointer support in the function, should you accept both here?

File: gremlin-go/driver/graphBinary.go [hunk 5]
```diff
@@ -873,10 +877,51 @@ func enumReader(buffer *bytes.Buffer, _ *graphBinaryTypeSerializer) (interface{}
...
+	b.Value = newValue
+
+	return b, nil
+}
+
 // gets the type of the serializer based on the value
 func (serializer *graphBinaryTypeSerializer) getSerializerToWrite(val interface{}) (*graphBinaryTypeSerializer, error) {
 	switch val.(type) {
-	case bytecode, *GraphTraversal:
+	case *bytecode, bytecode, *GraphTraversal:
```
Comment (lyndonbauto) [inline gremlin-go/driver/graphBinary.go:923]: Why is this required?

File: gremlin-go/driver/graphBinary.go (hunk 4 above)
Comment (lyndonbauto) [reply gremlin-go/driver/graphBinary.go]: Or remove support for non-pointers in the function?

File: gremlin-go/driver/request.go [hunk 6]
```diff
@@ -34,7 +34,8 @@ const sessionProcessor = "session"
 const stringOp = "eval"
 const stringProcessor = ""
 
-func makeStringRequest(stringGremlin string, traversalSource string, sessionId string) (req request) {
+// bindings should be a string key : object map (different from Binding class in bytecode)
```
Comment (lyndonbauto) [inline gremlin-go/driver/request.go]: Proper grammar in comments

File: gremlin-go/driver/bytecode.go [hunk 7]
```diff
@@ -152,8 +164,67 @@ type instruction struct {
...
+}
+
+// String returns the key value binding in string format
+func (b *Binding) String() string {
+	return fmt.Sprintf("binding[%v=%v]", b.Key, b.Value)
+}
+
+// Equals check for equality of current binding object to another
+func (b *Binding) Equals(item interface{}) bool {
+	if reflect.DeepEqual(b, item) {
```
Comment (lyndonbauto) [inline gremlin-go/driver/bytecode.go]: Should we be using DeepEquals? I think we found that it was very error prone and inconsistent.

File: gremlin-go/driver/bytecode.go [hunk 8]
```diff
@@ -152,8 +164,67 @@ type instruction struct {
...
+	}
+	if (item == nil) || fmt.Sprintf("%T", b) != fmt.Sprintf("%T", item) {
+		return false
+	}
+	val, ok := item.(*Binding)
+	if ok {
+		return b.Key == val.Key && reflect.DeepEqual(b.Value, val.Value)
+	} else {
+		return false
+	}
```
Comment (lyndonbauto) [inline gremlin-go/driver/bytecode.go]:
```suggestion
	return ok && (b.Key == val.Key && reflect.DeepEquals(b.Value, val.Value))
```

Comment (lyndonbauto) [reply gremlin-go/driver/bytecode.go]: Also same comment about deep equals.

File: gremlin-go/driver/bytecode.go [hunk 9]
```diff
@@ -122,16 +126,24 @@ func (bytecode *bytecode) convertArgument(arg interface{}) (interface{}, error)
...
-			bytecode.bindings[v.key] = v.value
-			return binding{
-				key:   v.key,
-				value: convertedValue,
+			bytecode.bindings[v.Key] = v.Value
+			return &Binding{
+				Key:   v.Key,
+				Value: convertedValue,
 			}, nil
+		case *GraphTraversal:
```
Comment (lyndonbauto) [inline gremlin-go/driver/bytecode.go:136]: Can we combine this and the Traversal case?

File: gremlin-go/driver/connection_test.go (hunk 1 above)
Comment (xiazcy) [reply gremlin-go/driver/connection_test.go]: Oh right, I was testing it locally and forgot to change it back.

File: gremlin-go/driver/graphBinary.go (hunk 2 above)
Comment (xiazcy) [reply gremlin-go/driver/graphBinary.go]: Looks like a mix-up from my merge attempt. Will delete.

File: gremlin-go/driver/graphBinary.go (hunk 3 above)
Comment (xiazcy) [reply gremlin-go/driver/graphBinary.go]: Again, a miss from the merge attempt.

File: gremlin-go/driver/request.go (hunk 6 above)
Comment (xiazcy) [reply gremlin-go/driver/request.go]: ACK.

File: gremlin-go/driver/graphBinary.go (hunk 5 above)
Comment (xiazcy) [reply gremlin-go/driver/graphBinary.go:923]: In the `convertArgument` function, Traversal/GraphTraversals has `*bytecode` in its struct, so it returns the pointer. It was only the default case that returned a `bytecode`, and we actually was not hitting the `Traversal` case previously, because our traversals are returned as pointers(`*GraphTraversal`), so `bytecode` serialization was enough, but with bindings I need the case that returns the `*bytecode`.

File: gremlin-go/driver/bytecode.go (hunk 9 above)
Comment (xiazcy) [reply gremlin-go/driver/bytecode.go:136]: ~~Actually yes, I'll need to change some test cases in `bytecode`, but they can both be under `*Traversal`~~
Maybe not, I'll have to see.

File: gremlin-go/driver/bytecode.go (hunk 7 above)
Comment (xiazcy) [reply gremlin-go/driver/bytecode.go]: Hmmm, I could look for a library that can be used? I'm not sure if I want to do what we did with the godog comparisons. This function isn't used extensively either.

File: gremlin-go/driver/bytecode.go (hunk 8 above)
Comment (xiazcy) [reply gremlin-go/driver/bytecode.go]: Changing this locally

File: gremlin-go/driver/graphBinary.go (hunk 4 above)
Comment (xiazcy) [reply gremlin-go/driver/graphBinary.go]: What do you mean here?

File: .github/workflows/build-test.yml [hunk 10]
```diff
@@ -14,194 +14,194 @@ jobs:
...
+#      - uses: actions/checkout@v2
+#      - name: Set up JDK11
+#        uses: actions/setup-java@v2
+#        with:
+#          java-version: '11'
+#          distribution: 'temurin'
+#      - name: Build with Maven
+#        run: |
+#          mvn clean install -pl -:gremlin-javascript,-:gremlin-python,-gremlin-dotnet,-:gremlin-dotnet-source,-:gremlin-dotnet-tests -q -DskipTests -Dci
+#          mvn verify -pl :neo4j-gremlin -DincludeNeo4j
```
Comment (L0Lmaker) [inline .github/workflows/build-test.yml:204]: Should we just remove them instead of commenting?

File: gremlin-go/driver/bytecode.go [hunk 11]
```diff
@@ -122,16 +126,24 @@ func (bytecode *bytecode) convertArgument(arg interface{}) (interface{}, error)
...
 			}
-			bytecode.bindings[v.key] = v.value
-			return binding{
-				key:   v.key,
-				value: convertedValue,
+			bytecode.bindings[v.Key] = v.Value
+			return &Binding{
+				Key:   v.Key,
+				Value: convertedValue,
 			}, nil
```
Comment (L0Lmaker) [inline gremlin-go/driver/bytecode.go:135]:
```suggestion
			return newBinding(v.Key, convertedValue), nil
```

File: gremlin-go/driver/bytecode.go [hunk 12]
```diff
@@ -152,8 +164,67 @@ type instruction struct {
...
+// will be reused over and over again save that some parameterization may change.
+// Used as g.V().Out(&Bindings{}.Of("key", value))
+type Bindings struct{}
+
+// Of creates a Binding
+func (*Bindings) Of(key string, value interface{}) *Binding {
+	return &Binding{
+		Key:   key,
+		Value: value,
+	}
```
Comment (L0Lmaker) [inline gremlin-go/driver/bytecode.go:178]:
```suggestion
	return newBinding(key, value)
```

File: gremlin-go/driver/bytecode_test.go [hunk 13]
```diff
@@ -108,15 +108,47 @@ func TestBytecode(t *testing.T) {
 		t.Run("binding", func(t *testing.T) {
 			testKey := "testKey"
 			testValue := "testValue"
-			testBinding := binding{
-				key:   testKey,
-				value: testValue,
+			testBinding := &Binding{
+				Key:   testKey,
+				Value: testValue,
```
Comment (L0Lmaker) [inline gremlin-go/driver/bytecode_test.go:113]:
```suggestion
			testBinding := newBinding(testKey, testValue)
```

File: gremlin-go/driver/bytecode_test.go [hunk 14]
```diff
@@ -108,15 +108,47 @@ func TestBytecode(t *testing.T) {
...
 			assert.Equal(t, testValue, bc.bindings[testKey])
 		})
+
+		t.Run("binding assert false", func(t *testing.T) {
+			testKey := "testKey"
+			testValue := 1
+			testBinding := &Binding{
+				Key:   testKey,
+				Value: testValue,
+			}
```
Comment (L0Lmaker) [inline gremlin-go/driver/bytecode_test.go]:
```suggestion
			testBinding := newBinding(testKey, testValue)
```

File: gremlin-go/driver/bytecode_test.go [hunk 15]
```diff
@@ -108,15 +108,47 @@ func TestBytecode(t *testing.T) {
...
+			converted, err := bc.convertArgument(testBinding)
+			assert.Nil(t, err)
+			assert.Equal(t, testBinding, converted)
+			assert.Equal(t, testValue, bc.bindings[testKey])
+			assert.False(t, converted.(*Binding).Equals(&Binding{
+				Key:   "t",
+				Value: "v",
+			}))
```
Comment (L0Lmaker) [inline gremlin-go/driver/bytecode_test.go]:
```suggestion
			assert.False(t, converted.(*Binding).Equals(newBinding("t", "v")))
```

File: gremlin-go/driver/bytecode_test.go [hunk 16]
```diff
@@ -108,15 +108,47 @@ func TestBytecode(t *testing.T) {
...
+		})
+
+		t.Run("binding hash", func(t *testing.T) {
+			testKey := "testKey"
+			// due to strict typing, we are converting all values of slice/map arguments into interface{}
+			testValue := []interface{}{1}
+			testBinding := &Binding{
+				Key:   testKey,
+				Value: testValue,
+			}
```
Comment (L0Lmaker) [inline gremlin-go/driver/bytecode_test.go]:
```suggestion
			testBinding := newBinding(testKey, testValue)
```

File: gremlin-go/driver/graphBinary_test.go [hunk 17]
```diff
@@ -260,6 +260,14 @@ func TestGraphBinaryV1(t *testing.T) {
 			e := []interface{}{int32(1), int32(1), int64(3), int64(3), int64(3)}
 			assert.Equal(t, e, readToValue(&buff))
 		})
+		t.Run("test Binding", func(t *testing.T) {
+			x := &Binding{
+				Key:   "key",
+				Value: "value",
+			}
```
Comment (L0Lmaker) [inline gremlin-go/driver/graphBinary_test.go:267]:
```suggestion
			x := newBinding("key", "value")
```

File: gremlin-go/driver/traversal_test.go [hunk 18]
```diff
@@ -43,4 +43,27 @@ func TestTraversal(t *testing.T) {
...
+		assert.Equal(t, 5, len(bytecode.bindings))
+		assert.Equal(t, []int32{1, 2, 3}, bytecode.bindings["a"])
+		assert.Equal(t, "created", bytecode.bindings["b"])
+		assert.Equal(t, "created", bytecode.bindings["c"])
+		assert.Equal(t, "knows", bytecode.bindings["d"])
+		assert.Equal(t, P.Gt(2), bytecode.bindings["e"])
+		assert.Equal(t, &Binding{
+			Key:   "b",
+			Value: "created",
+		}, bytecode.stepInstructions[1].arguments[0])
```
Comment (L0Lmaker) [inline gremlin-go/driver/traversal_test.go:62]:
```suggestion
		assert.Equal(t, newBinding("b", "created"), bytecode.stepInstructions[1].arguments[0])
```

File: gremlin-go/driver/traversal_test.go [hunk 19]
```diff
@@ -43,4 +43,27 @@ func TestTraversal(t *testing.T) {
...
+		assert.True(t, bytecode.stepInstructions[1].arguments[0].(*Binding).Equals(&Binding{
+			Key:   "b",
+			Value: "created",
+		}))
```
Comment (L0Lmaker) [inline gremlin-go/driver/traversal_test.go]:
```suggestion
		assert.True(t, bytecode.stepInstructions[1].arguments[0].(*Binding).Equals(newBinding("b", "created")))
```

File: .github/workflows/build-test.yml (hunk 10 above)
Comment (xiazcy) [reply .github/workflows/build-test.yml:204]: Maybe I should. It'll look cleaner with it deleted.

File: gremlin-go/driver/bytecode.go (hunk 11 above)
Comment (xiazcy) [reply gremlin-go/driver/bytecode.go:135]: Ah, I'm actually not sure if I really needed the `newBinding` constructor, since we can create Bindings via structs, it was left from an attempt to set up the `Bindings.Of` call differently. Perhaps this can be discussed.

File: gremlin-go/driver/bytecode.go [hunk 20]
```diff
@@ -152,8 +164,59 @@ type instruction struct {
...
+	err = enc.Encode(b.Value)
+	if err != nil {
+		return 0
+	}
+	valHash, err := h.Write(buff.Bytes())
+	if err != nil {
+		return 0
+	}
+	return keyHash + valHash
+}
```
Comment (lyndonbauto) [inline gremlin-go/driver/bytecode.go]: Remove Equals and Hash. This is python junk.
```suggestion
```

File: gremlin-go/driver/bytecode.go (hunk 9 above)
Comment (xiazcy) [reply gremlin-go/driver/bytecode.go:136]: I'm thinking of removing the `Traversal` case entirely, as we actually don't return `Traversal`/`*Traversal` at any point, just `*GraphTraversals`.

File: gremlin-go/driver/bytecode.go (hunk 20 above)
Comment (xiazcy) [reply gremlin-go/driver/bytecode.go]: Will remove locally

**Review by lyndonbauto — APPROVED** [review]
